synapse/state/v2.py

```diff
@@ -252,9 +252,76 @@ async def _get_auth_chain_difference(
         Set of event IDs
     """
 
-    difference = await state_res_store.get_auth_chain_difference(
-        [set(state_set.values()) for state_set in state_sets]
-    )
+    # The `StateResolutionStore.get_auth_chain_difference` function assumes that
+    # all events passed to it (and their auth chains) have been persisted
+    # previously. This is not the case for any events in the `event_map`, and so
+    # we need to manually handle those events.
+    #
+    # We do this by calculating the auth chain difference based on events in
+    # `event_map` and adding that to the result from calling
+    # `get_auth_chain_difference` with state sets where we've replaced
+    # references to events in `event_map` with their auth events (recursively).
+
+    # Map from event ID in `event_map` to their auth event IDs, and their auth
+    # event IDs if they appear in the `event_map`. This is the intersection of
+    # the event's auth chain with the events in the `event_map` *plus* their
+    # auth event IDs.
+    events_to_auth_chain = {}
+    for event in event_map.values():
+        chain = {event.event_id}
+        events_to_auth_chain[event.event_id] = chain
+
+        to_search = [event]
+        while to_search:
+            for auth_id in to_search.pop().auth_event_ids():
+                chain.add(auth_id)
+                auth_event = event_map.get(auth_id)
+                if auth_event:
+                    to_search.append(auth_event)
+
+    # We now a) calculate the auth chain difference for the unpersisted events
+    # and b) work out the state sets to pass to the store.
+    #
+    # Note: If the `event_map` is empty (which is the common case), we can do a
+    # much simpler calculation.
```

Comment:
you may as well stick the `events_to_auth_chain` logic inside the `if event_map` condition too.

Reply:
I found it harder to follow with all the extra indents TBH. I don't think it has much of a performance impact since its just a case of creating an empty dict and iterating over an empty dict.

Reply:
yeah I'm not worried about the performance impact. It just feels odd to me that we've decided we're going to optimise *part* of the algorithm for the "`event_map` is empty" case, and not the other half - for me that makes it harder to follow.

I don't feel strongly though: up to you.